### ISSUE TITLE
Document overheads and guarantees?

### ISSUE THREAD

**Issue by rbtcollins** [libpnet/libpnet#190]
I was looking to do packet manipulation on top of netmap_sys (specifically looking for 10+Gbps line-rate capabilities, so DPDK and netmap and PF_RING_ZC are the only options AIUI), and pnet looks like an interesting possiblity, but it seems - at brief inspection - to introduce a number of inefficiencies on top of the basic batch operation. I'm sure I'm not getting all the subtleties of rust though, so would be great would be something in the docs / README that talks about overheads.

E.g. "libpnet supports zero-copy receive and send of packets with DPDK/netmap - e.g. a router receiving packets, updating the destination MAC and sending on the same interface can do so with no copies involved." [or whatever the actual capabilities of pnet are].

**Comment by mrmonday** [comment]
Hi there,

When using the netmap backend libpnet has no problem saturating a 10Gbit link (in fact, in previous benchmarks I've run it outperforms the equivalent C code on the same hardware - that's likely just down to different sets of compiler optimisations though).

May I ask what inefficiencies you can see? Perhaps I could address them if you are more specific. Receiving definitely isn't optimal currently, but it is fast enough for a 10Gbit link. There may be issues on bigger links, unfortunately I've never had access to >10Gbit hardware to test. There should be no overhead when using libpnet compared to writing the equivalent C which netmap directly.

This paper might be of interest if you're looking for more information: https://octarineparrot.com/assets/msci_paper.pdf. It was written with a deadline, so please ask if anything's unclear. Some of it is a little out of date too now.

Hope this clears a few things up for you, let me know if you have any more questions.

**Comment by rbtcollins** [comment]
I haven't got any benchmarks or anything as yet, the main thing was that the docs didn't give me any clue about pnet's performance guarantees/overheads - leaving me to make (apparently poor) conclusions from quick reading of code :/.  As a for-instance, the send iterator that takes a callback and supplies a mutable buffer - that appears to require a copy.

**Comment by mrmonday** [comment]
The documentation should be improved, yes.

The callback takes a `MutableEthernetPacket`, which is just a `&mut [u8]`, that is, a mutable slice - which is represented as a pointer and length pair (https://doc.rust-lang.org/1.9.0/std/raw/struct.Slice.html). This allows the callback to write directly into netmap's buffers without any copying. In the applications I've written, I've found that the callback is usually inlined too, removing the function call overhead.

The API is specifically designed to have little or no overhead compared to using the unsafe netmap API directly.

**Comment by rbtcollins** [comment]
More details:
The example here - https://octarineparrot.com/assets/libpnet/doc/pnet/index.html#ethernet-echo-server

says it builds the packet in the write buffer with no copy, but clone_from does a copy:https://github.com/libpnet/libpnet/blob/a5f9c50011321e625179ecee926cda037aa0bc57/src/packet/mod.rs#L31, and since the receive iterator returns EthernetPacket, not MutableEthernetPacket, it seems to me that the copy is baked into the API?

**Comment by mrmonday** [comment]
That specific example does do a copy, yes.

The packet is copied directly from one netmap buffer to another. There's no support for netmap's zero-copy packet forwarding currently, I've opened #192 to track that (pull requests welcome! :smiley:).

Are you looking to do software switching with libpnet? I'd be happy to help design and implement the API for this if it's something you're interested in working on. It shouldn't be too complex to implement.

If you're doing protocol implementation or traffic analysis it can be done with zero copies with the existing API (depending on the protocol you're implementing of course).

**Comment by rbtcollins** [comment]
Yes, its a form of switching I'm poking around at - based on / inspired by Google's maglev paper. I've given more details in #192; lets leave this ticket focused on the docs?
